// @arbitrum/sdk-nitro/dist/lib/message/L1Transaction.js
/*
 * Copyright 2021, Offchain Labs, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/* eslint-env node */
'use strict';
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.L1ContractCallTransactionReceipt = exports.L1EthDepositTransactionReceipt = exports.L1TransactionReceipt = void 0;
const bignumber_1 = require("@ethersproject/bignumber");
const L1ToL2Message_1 = require("./L1ToL2Message");
const L1ERC20Gateway__factory_1 = require("../abi/factories/L1ERC20Gateway__factory");
const signerOrProvider_1 = require("../dataEntities/signerOrProvider");
const errors_1 = require("../dataEntities/errors");
const ethers_1 = require("ethers");
const Inbox__factory_1 = require("../abi/factories/Inbox__factory");
const bytes_1 = require("@ethersproject/bytes");
const message_1 = require("../dataEntities/message");
const Bridge__factory_1 = require("../abi/factories/Bridge__factory");
class L1TransactionReceipt {
    constructor(tx) {
        this.to = tx.to;
        this.from = tx.from;
        this.contractAddress = tx.contractAddress;
        this.transactionIndex = tx.transactionIndex;
        this.root = tx.root;
        this.gasUsed = tx.gasUsed;
        this.logsBloom = tx.logsBloom;
        this.blockHash = tx.blockHash;
        this.transactionHash = tx.transactionHash;
        this.logs = tx.logs;
        this.blockNumber = tx.blockNumber;
        this.confirmations = tx.confirmations;
        this.cumulativeGasUsed = tx.cumulativeGasUsed;
        this.effectiveGasPrice = tx.effectiveGasPrice;
        this.byzantium = tx.byzantium;
        this.type = tx.type;
        this.status = tx.status;
    }
    /**
     * Get any MessageDelivered events that were emitted during this transaction
     * @returns
     */
    getMessageDeliveredEvents() {
        const iface = Bridge__factory_1.Bridge__factory.createInterface();
        const messageDeliveredTopic = iface.getEventTopic(iface.getEvent('MessageDelivered'));
        return this.logs
            .filter(log => log.topics[0] === messageDeliveredTopic)
            .map(l => iface.parseLog(l).args);
    }
    /**
     * Get any InboxMessageDelivered events that were emitted during this transaction
     * @returns
     */
    getInboxMessageDeliveredEvent() {
        const iFace = Inbox__factory_1.Inbox__factory.createInterface();
        const inboxMessageDeliveredTopic = iFace.getEventTopic(iFace.events['InboxMessageDelivered(uint256,bytes)']);
        return this.logs
            .filter(log => log.topics[0] === inboxMessageDeliveredTopic)
            .map(l => iFace.parseLog(l).args);
    }
    /**
     * Get combined data for any InboxMessageDelivered and MessageDelivered events
     * emitted during this transaction
     * @returns
     */
    getMessageEvents() {
        const bridgeMessages = this.getMessageDeliveredEvents();
        const inboxMessages = this.getInboxMessageDeliveredEvent();
        if (bridgeMessages.length !== inboxMessages.length) {
            throw new errors_1.ArbSdkError(`Unexpected missing events. Inbox message count: ${inboxMessages.length} does not equal bridge message count: ${bridgeMessages.length}. ${JSON.stringify(inboxMessages)} ${JSON.stringify(bridgeMessages)}`);
        }
        const messages = [];
        for (const bm of bridgeMessages) {
            const im = inboxMessages.filter(i => i.messageNum.eq(bm.messageIndex))[0];
            if (!im) {
                throw new errors_1.ArbSdkError(`Unexepected missing event for message index: ${bm.messageIndex.toString()}. ${JSON.stringify(inboxMessages)}`);
            }
            messages.push({
                inboxMessageEvent: im,
                bridgeMessageEvent: bm,
            });
        }
        return messages;
    }
    parseRetryableMessageData(eventData) {
        // decode the data field - is been packed so we cant decode the bytes field this way
        const parsed = ethers_1.ethers.utils.defaultAbiCoder.decode([
            'uint256',
            'uint256',
            'uint256',
            'uint256',
            'uint256',
            'uint256',
            'uint256',
            'uint256',
            'uint256', // data length
        ], 
        // decode from the first 9 words
        eventData.substring(0, 64 * 9 + 2));
        const addressFromBigNumber = (bn) => ethers_1.ethers.utils.getAddress((0, bytes_1.hexZeroPad)(bn.toHexString(), 20));
        const destAddress = addressFromBigNumber(parsed[0]);
        const l2CallValue = parsed[1];
        const l1Value = parsed[2];
        const maxSubmissionFee = parsed[3];
        const excessFeeRefundAddress = addressFromBigNumber(parsed[4]);
        const callValueRefundAddress = addressFromBigNumber(parsed[5]);
        const gasLimit = parsed[6];
        const maxFeePerGas = parsed[7];
        const data = '0x' + eventData.substring(64 * 9 + 2);
        return {
            destAddress,
            l2CallValue,
            l1Value,
            maxSubmissionFee: maxSubmissionFee,
            excessFeeRefundAddress,
            callValueRefundAddress,
            gasLimit,
            maxFeePerGas,
            data,
        };
    }
    async getEthDepositMessages(l2Provider) {
        return Promise.all(this.getMessageEvents()
            .filter(e => e.bridgeMessageEvent.kind ===
            message_1.InboxMessageKind.L1MessageType_ethDeposit)
            .map(m => L1ToL2Message_1.EthDepositMessage.fromEventComponents(l2Provider, m.inboxMessageEvent.messageNum, m.bridgeMessageEvent.sender, m.inboxMessageEvent.data)));
    }
    async getL1ToL2Messages(l2SignerOrProvider) {
        const provider = signerOrProvider_1.SignerProviderUtils.getProviderOrThrow(l2SignerOrProvider);
        const chainID = (await provider.getNetwork()).chainId.toString();
        const events = this.getMessageEvents();
        return events
            .filter(e => e.bridgeMessageEvent.kind ===
            message_1.InboxMessageKind.L1MessageType_submitRetryableTx)
            .map(mn => {
            const inboxMessageData = this.parseRetryableMessageData(mn.inboxMessageEvent.data);
            return L1ToL2Message_1.L1ToL2Message.fromTxComponents(l2SignerOrProvider, bignumber_1.BigNumber.from(chainID).toNumber(), mn.bridgeMessageEvent.sender, mn.inboxMessageEvent.messageNum, mn.bridgeMessageEvent.baseFeeL1, inboxMessageData);
        });
    }
    /**
     * Get any token deposit events created by this transaction
     * @returns
     */
    getTokenDepositEvents() {
        const iface = L1ERC20Gateway__factory_1.L1ERC20Gateway__factory.createInterface();
        const event = iface.getEvent('DepositInitiated');
        const eventTopic = iface.getEventTopic(event);
        const logs = this.logs.filter(log => log.topics[0] === eventTopic);
        return logs.map(log => iface.parseLog(log).args);
    }
}
exports.L1TransactionReceipt = L1TransactionReceipt;
_a = L1TransactionReceipt;
/**
 * Replaces the wait function with one that returns an L1TransactionReceipt
 * @param contractTransaction
 * @returns
 */
L1TransactionReceipt.monkeyPatchWait = (contractTransaction) => {
    const wait = contractTransaction.wait;
    contractTransaction.wait = async (confirmations) => {
        const result = await wait(confirmations);
        return new L1TransactionReceipt(result);
    };
    return contractTransaction;
};
/**
 * Replaces the wait function with one that returns an L1EthDepositTransactionReceipt
 * @param contractTransaction
 * @returns
 */
L1TransactionReceipt.monkeyPatchEthDepositWait = (contractTransaction) => {
    const wait = contractTransaction.wait;
    contractTransaction.wait = async (confirmations) => {
        const result = await wait(confirmations);
        return new L1EthDepositTransactionReceipt(result);
    };
    return contractTransaction;
};
/**
 * Replaces the wait function with one that returns an L1ContractCallTransactionReceipt
 * @param contractTransaction
 * @returns
 */
L1TransactionReceipt.monkeyPatchContractCallWait = (contractTransaction) => {
    const wait = contractTransaction.wait;
    contractTransaction.wait = async (confirmations) => {
        const result = await wait(confirmations);
        return new L1ContractCallTransactionReceipt(result);
    };
    return contractTransaction;
};
/**
 * An L1TransactionReceipt with additional functionality that only exists
 * if the transaction created a single eth deposit.
 */
class L1EthDepositTransactionReceipt extends L1TransactionReceipt {
    /**
     * Wait for the funds to arrive on L2
     * @param confirmations Amount of confirmations the retryable ticket and the auto redeem receipt should have
     * @param timeout Amount of time to wait for the retryable ticket to be created
     * Defaults to 15 minutes, as by this time all transactions are expected to be included on L2. Throws on timeout.
     * @returns The wait result contains `complete`, a `status`, the L1ToL2Message and optionally the `l2TxReceipt`
     * If `complete` is true then this message is in the terminal state.
     * For eth deposits complete this is when the status is FUNDS_DEPOSITED, EXPIRED or REDEEMED.
     */
    async waitForL2(l2Provider, confirmations, timeout = 900000) {
        const message = (await this.getEthDepositMessages(l2Provider))[0];
        if (!message)
            throw new errors_1.ArbSdkError('Unexpected missing Eth Deposit message.');
        const res = await message.wait(confirmations, timeout);
        return {
            complete: !!res,
            l2TxReceipt: res,
            message,
        };
    }
}
exports.L1EthDepositTransactionReceipt = L1EthDepositTransactionReceipt;
/**
 * An L1TransactionReceipt with additional functionality that only exists
 * if the transaction created a single call to an L2 contract - this includes
 * token deposits.
 */
class L1ContractCallTransactionReceipt extends L1TransactionReceipt {
    /**
     * Wait for the transaction to arrive and be executed on L2
     * @param confirmations Amount of confirmations the retryable ticket and the auto redeem receipt should have
     * @param timeout Amount of time to wait for the retryable ticket to be created
     * Defaults to 15 minutes, as by this time all transactions are expected to be included on L2. Throws on timeout.
     * @returns The wait result contains `complete`, a `status`, an L1ToL2Message and optionally the `l2TxReceipt`.
     * If `complete` is true then this message is in the terminal state.
     * For contract calls this is true only if the status is REDEEMED.
     */
    async waitForL2(l2SignerOrProvider, confirmations, timeout = 900000) {
        const message = (await this.getL1ToL2Messages(l2SignerOrProvider))[0];
        if (!message)
            throw new errors_1.ArbSdkError('Unexpected missing L1ToL2 message.');
        const res = await message.waitForStatus(confirmations, timeout);
        return Object.assign(Object.assign({ complete: res.status === L1ToL2Message_1.L1ToL2MessageStatus.REDEEMED }, res), { message });
    }
}
exports.L1ContractCallTransactionReceipt = L1ContractCallTransactionReceipt;
