// @arbitrum/sdk/dist/lib/utils/arbProvider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getArbBlockByHash = exports.getArbTransactionReceipt = void 0;
const providers_1 = require("@ethersproject/providers");
const constants_1 = require("../dataEntities/constants");
const NodeInterface__factory_1 = require("../abi/factories/NodeInterface__factory");
class ArbFormatter extends providers_1.Formatter {
    getDefaultFormats() {
        // formats was already initialised in super, so we can just access here
        const superFormats = super.getDefaultFormats();
        const bigNumber = this.bigNumber.bind(this);
        const hash = this.hash.bind(this);
        const number = this.number.bind(this);
        const arbBlockProps = {
            sendRoot: hash,
            sendCount: bigNumber,
            l1BlockNumber: number,
        };
        const arbReceiptFormat = Object.assign(Object.assign({}, superFormats.receipt), { l1BlockNumber: number, gasUsedForL1: bigNumber });
        return Object.assign(Object.assign({}, superFormats), { receipt: arbReceiptFormat, block: Object.assign(Object.assign({}, superFormats.block), arbBlockProps), blockWithTransactions: Object.assign(Object.assign({}, superFormats.blockWithTransactions), arbBlockProps) });
    }
    receipt(value) {
        return super.receipt(value);
    }
    block(block) {
        return super.block(block);
    }
    blockWithTransactions(block) {
        return super.blockWithTransactions(block);
    }
}
async function getArbTransactionReceipt(l2Provider, txHash, fetchBatchNumber, fetchBatchConfirmations) {
    const rec = await l2Provider.send('eth_getTransactionReceipt', [txHash]);
    if (rec == null)
        return null;
    const arbFormatter = new ArbFormatter();
    const arbTxReceipt = arbFormatter.receipt(rec);
    const nodeInterface = NodeInterface__factory_1.NodeInterface__factory.connect(constants_1.NODE_INTERFACE_ADDRESS, l2Provider);
    if (fetchBatchNumber) {
        // findBatchContainingBlock errors if block number does not exist
        try {
            const res = await nodeInterface.findBatchContainingBlock(arbTxReceipt.blockNumber);
            arbTxReceipt.l1BatchNumber = res.toNumber();
        }
        catch (err) {
            // do nothing - errors are expected here
        }
    }
    if (fetchBatchConfirmations) {
        // getL1Confirmations returns 0 if block has does not exist
        const res = await nodeInterface.getL1Confirmations(arbTxReceipt.blockHash);
        arbTxReceipt.l1BatchConfirmations = res.toNumber();
    }
    return arbTxReceipt;
}
exports.getArbTransactionReceipt = getArbTransactionReceipt;
async function getArbBlockByHash(l2Provider, blockHash, includeTransactions) {
    const l2Block = await l2Provider.send('eth_getBlockByHash', [
        blockHash,
        includeTransactions,
    ]);
    if (l2Block === null)
        return null;
    const arbFormatter = new ArbFormatter();
    return includeTransactions
        ? arbFormatter.blockWithTransactions(l2Block)
        : arbFormatter.block(l2Block);
}
exports.getArbBlockByHash = getArbBlockByHash;
